### PR TITLE
fix(cli): hide diffs of mangled unicode strings

### LINKED ISSUE #25309: (cdk-core): CDK diff does not work with umlauts
### Describe the bug

When doing a CDK diff it always detect a change in the tags because the „ü“ in this case results in a „?“. 
Although the Tags on the resource are correct including the umlaut as specified.

### Expected Behavior

Diff command should be capable to handle all allowed chars for tags.

### Current Behavior

[~] AWS::IAM::Role VirtualGuideiaaInstance/Cert/Certificate/CertificateRequestorFunction/ServiceRole VirtualGuideiaaInstanceCertCertificateCertificateRequestorFunctionServiceRole3B312417 
 └─ [~] Tags
     └─ @@ -5,7 +5,7 @@
        [ ] },
        [ ] {
        [ ]   "Key": "customer",
        [-]   "Value": "Messe M?nchen"
        [+]   "Value": "Messe München"
        [ ] },
        [ ] {
        [ ]   "Key": "environment",

### Reproduction Steps

const app = new App();
// ...
Tags.of(app).add("Ära", "München")

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### CDK CLI Version

2.76.0 (build 78c411b)

### Framework Version

2.76.0

### Node.js Version

v16.17.0

### OS

Mac OS 13.3

### Language

Typescript

### Language Version

3.9.10

### Other information

_No response_

### PR DESCRIPTION
**Pull request by laverdet**
I am reopening this from https://github.com/aws/aws-cdk/pull/25008

> This PR has been deemed to be abandoned, and will be automatically closed. Please create a new PR for these changes if you think this decision has been made in error.

and following up on my comments here:
https://github.com/aws/aws-cdk/pull/24557#issuecomment-1492965718
https://github.com/aws/aws-cdk/pull/24557#issuecomment-1500747296
https://github.com/aws/aws-cdk/pull/25008#issuecomment-1501441939
https://github.com/aws/aws-cdk/pull/25008#issuecomment-1518734007
https://github.com/aws/aws-cdk/pull/25008#issuecomment-1526860767
https://github.com/aws/aws-cdk/pull/25008#issuecomment-1530705196
https://github.com/aws/aws-cdk/pull/25008#issuecomment-1532033381
https://github.com/aws/aws-cdk/pull/25008#issuecomment-1541047453

@aws-cdk-automation @rix0rrr @TheRealAmazonKendra - I'm happy to adjust the approach, add more tests, or do what else needs to be done. I'm not getting any feedback from the team so I'm not sure how to proceed. The diff noise with non-ASCII information in cdk diff makes it difficult to find meaningful changes to our stacks.

🗿🗞️📬 **Crucially, this change only affects the CLI output and therefore an integration test isn't possible.**

---

CloudFormation's `GetStackTemplate` irrecoverably mangles any character not in the 7-bit ASCII range. This causes noisy output from `cdk diff` when a template contains non-English languages or emoji. We can detect this case and consider these strings equal.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

Many AWS services accept non-ASCII input, eg many "description" fields. CloudFormation will correctly dispatch these templates but when invoking `GetStackTemplate` the result is mangled. This causes annoying noise in the output of `cdk diff`:

```
Resources
[~] AWS::Lambda::Function Lambda/Resource
 └─ [~] Description
     ├─ [-] ?????
     └─ [+] 🤦🏻‍♂️
```

This change modifies the diff algorithm to consider the string equal if the lvalue is a mangled version of the rvalue.

Of course this runs the risk of hiding changesets which modify only a single non-ASCII character to another non-ASCII character, but these fields already tend to be informative in nature.

fixes #25309

### CODE REVIEW

**Comment by laverdet** [comment]
Exemption Request - This change only affects the CLI output and therefore an integration test isn't possible. I'm happy to spend more time documenting or working through other prerequisites but I do want to make sure that, conceptually, this change is one that the team will accept. I have also included more information about the benefits and tradeoffs of this change in the original PR message.

**Comment by laverdet** [comment]
@skinny85 any chance I could get some eyes on this?

**Comment by skinny85** [comment]
@laverdet unfortunately, I'm no longer on the CDK team, so I can't help with the PR 🙂.

**Comment by laverdet** [comment]
@corymhall Is there any chance you or someone on the team could take a quick look at this? It is a 3-line CLI-only QOL improvement and I'm a little confused as to why I haven't gotten any communication from the team in the 2 months since I opened #24557. If this is not something the team wants then that's ok but right now I'm caught in an awkward holding pattern.

File: packages/@aws-cdk/cloudformation-diff/test/deep-equal.test.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { deepEqual } from '../lib/diff/util';
+
+test('mangled strings', () => {
+  expect(deepEqual('foo', 'foo')).toBeTruthy();
+  expect(deepEqual('????', '文字化け')).toBeTruthy();
```
Comment (otaviomacedo) [inline packages/@aws-cdk/cloudformation-diff/test/deep-equal.test.ts:5]: Equality, being an equivalence relation, should be reflexive, symmetric and transitive. In particular, for the symmetric property, the following should hold:

`deepEqual('????', '文字化け')` $\iff$ `deepEqual('文字化け', '????')`

which is not the case. The left-hand side is true and the right-hand side is false. But even if we fix this (and assuming transitivity also holds), we get the following nonsensical implication:

```ts
// Given that the following statements are true:
deepEqual('文字化け', '????')
deepEqual('????', '🤦🏻‍♂️')

// Then, by transitivity:
deepEqual('文字化け', '🤦🏻‍♂️')
```

**Comment by laverdet** [comment]
Please don't close it again. It is clearly not abandoned.

**Pull request closed without merging**